lib/ProductOpener/Ingredients.pm

```diff
@@ -886,6 +887,60 @@ sub add_properties_from_specific_ingredients ($product_ref) {
 	return;
 }
 
+=head2 add_percent_max_for_salt_and_sugar_ingredients_from_nutrition_facts ( $product_ref )
+
+Add a percent_max value for salt and sugar ingredients, based on the nutrition facts.
+
+=cut
+
+sub add_percent_max_for_salt_and_sugar_ingredients_from_nutrition_facts ($product_ref) {
+
+	# Check if we have values for salt and sugar in the nutrition facts
+	my @ingredient_max_values = ();
+	my $sugars_100g = deep_get($product_ref, qw(nutriments sugars_100g));
+	if (defined $sugars_100g) {
+		push @ingredient_max_values, {ingredientid => "en:sugar", value => $sugars_100g};
+	}
+	my $salt_100g = deep_get($product_ref, qw(nutriments salt_100g));
+	if (defined $salt_100g) {
+		push @ingredient_max_values, {ingredientid => "en:salt", value => $salt_100g};
+	}
+
+	if ((defined $sugars_100g) or (defined $salt_100g)) {
+
+		foreach my $ingredient_max_value_ref (@ingredient_max_values) {
+
+			# Traverse the ingredients tree, breadth first
+
+			my @ingredients = @{$product_ref->{ingredients}};
+
+			while (@ingredients) {
+
+				# Remove and process the first ingredient
+				my $ingredient_ref = shift @ingredients;
+				my $ingredientid = $ingredient_ref->{id};
+
+				# Add sub-ingredients at the beginning of the ingredients array
+				if (defined $ingredient_ref->{ingredients}) {
+
+					unshift @ingredients, @{$ingredient_ref->{ingredients}};
+				}
+
+				foreach my $ingredient_max_value_ref (@ingredient_max_values) {
```

Comment:
We already iterated above ! Why do we need to iterate twice ?
(I prefer if we keep this inner loop, that said).

Reply:
ah good point, initially I had it in the inner loop but I wanted to move it to the outer loop, and I forgot to remove the inner loop.. :(

we can keep the inner loop only